isabelle: 2020 -> 2021

meithecatte opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

meithecatte committed:
naproche: init at 0.1.0.0

meithecatte committed:
polyml: Use make compiler instead of plain make

'make compiler' results in a fuller bootstrap, such that the behavior of
the resulting binary is consistent with the code. In the Git commit of
PolyML used by Isabelle, using just 'make' manifested by the contents of
the Foreign module going missing.

meithecatte committed:
isabelle: 2020 -> 2021

meithecatte committed:
isabelle: build native SHA1 implementation

This eliminates the warnings about a slow fallback implementation being
used that could be seen in heap build logs

Based on code by Nick Spinale

meithecatte committed:
isabelle: build system heaps

Isabelle ships pre-built a heap for Pure and HOL. However, due to the
changes introduced by nixpkgs, they aren't considered up to date.
Rebuild the heaps, so that the users don't have to do so on launch.

r-rmcgibbo commented:
Result of `nixpkgs-review pr 115339` at 86f31bb6 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages marked as broken and skipped:</summary>
  <ul>
    <li>isabelle</li>
    <li>tlaps</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li><a href="https://gist.github.com/f23983dfa100f5b5123a0489ae924e67">naproche</a></li></li>
    <li><a href="https://gist.github.com/249a07dc6115065fbcab113f4350fb44">polyml</a></li></li>
  </ul>
</details>
<details>
  <summary>3 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-phase-hooks.md">missing-phase-hooks</a>

`buildPhase` should probably contain `runHook preBuild` and `runHook postBuild`.

Near pkgs/development/compilers/polyml/default.nix:21:3:
```
   |
21 |   buildPhase = ''
   |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`lgpl21` is a deprecated license, check if project uses `lgpl21Plus` or `lgpl21Only` and change `meta.license` accordingly.

Near pkgs/development/compilers/polyml/default.nix:39:5:
```
   |
39 |     license = licenses.lgpl21;
   |     ^
```
</li>
    <li>warning: unused-argument

Unused argument: `fetchpatch`.
Near pkgs/development/compilers/polyml/default.nix:1:33:
```
  |
1 | { lib, stdenv, fetchFromGitHub, fetchpatch, autoreconfHook, gmp, libffi }:
  |                                 ^
```
</li>
  </ul>
</details>

SuperSandro2000 commented on pkgs/applications/science/logic/isabelle/default.nix:
```diff
@@ -58,8 +72,21 @@ stdenv.mkDerivation rec {
+    substituteInPlace lib/Tools/env \
+      --replace /usr/bin/env ${coreutils}/bin/env
+
+    rm -rf lib/classes heaps
```
```suggestion
    rm -r lib/classes heaps
```

SuperSandro2000 commented on pkgs/applications/science/logic/isabelle/sha1.nix:
```diff
@@ -0,0 +1,27 @@
+{ stdenv, lib, fetchhg }:
+
+stdenv.mkDerivation {
+  name = "isabelle-sha1";
```
This should probably also use pname + version.

NieDzejkob replied:
AFAICS this component has no concept of version.

SuperSandro2000 commented on pkgs/applications/science/logic/isabelle/sha1.nix:
```diff
@@ -0,0 +1,27 @@
+    mkdir -p $out/lib
+    cp libsha1.so $out/lib/libsha1.so
+  '';
+
```
Little meta section wouldn't hurt.

SuperSandro2000 commented on pkgs/applications/science/logic/naproche/default.nix:
```diff
@@ -0,0 +1,26 @@
+{ lib, fetchFromGitHub, haskellPackages }:
+with haskellPackages; mkDerivation {
```
```suggestion
{ lib, fetchFromGitHub, haskellPackages }:

with haskellPackages; mkDerivation {
```

SuperSandro2000 commented on pkgs/applications/science/logic/naproche/default.nix:
```diff
@@ -0,0 +1,26 @@
+{ lib, fetchFromGitHub, haskellPackages }:
+with haskellPackages; mkDerivation {
+  pname = "Naproche-SAD";
+  version = "0.1.0.0";
+  src = fetchFromGitHub {
```
```suggestion
  version = "0.1.0.0";

  src = fetchFromGitHub {
```

SuperSandro2000 commented on pkgs/applications/science/logic/naproche/default.nix:
```diff
@@ -0,0 +1,26 @@
+    # This revision is used by Isabelle
+    rev = "755224402e36bf8276c9e3d5a8eeb024bb911b85";
+    sha256 = "0zlxj10iksnn3j295b91n23y9cxwxa8d95c326karsv59bpzvaz3";
+  };
+  isExecutable = true;
```
```suggestion
  };

  isExecutable = true;
```

SuperSandro2000 commented on pkgs/applications/science/logic/naproche/default.nix:
```diff
@@ -0,0 +1,26 @@
+  src = fetchFromGitHub {
+    owner = "naproche-community";
+    repo = "naproche";
+    # This revision is used by Isabelle
```
Is this strictly required or just some arbitrary pin?

NieDzejkob replied:
No released version is new enough, by a large margin. It seems that development was taken over by the Isabelle team, and now the closest thing to a release is the commit used by a release of Isabelle.

SuperSandro2000 commented on pkgs/applications/science/logic/naproche/default.nix:
```diff
@@ -0,0 +1,26 @@
+  isExecutable = true;
+  # test suite requires 'stack exec'
+  doCheck = false;
+  libraryToolDepends = [ hpack ];
+  executableHaskellDepends = [
```
```suggestion
  libraryToolDepends = [ hpack ];

  executableHaskellDepends = [
```

SuperSandro2000 commented on pkgs/applications/science/logic/naproche/default.nix:
```diff
@@ -0,0 +1,26 @@
+  executableHaskellDepends = [
+    base bytestring containers ghc-prim mtl network process split text
+    threads time transformers utf8-string uuid yaml
+  ];
+  prePatch = "hpack";
```
```suggestion
  ];

  prePatch = "hpack";
```

SuperSandro2000 commented on pkgs/development/compilers/polyml/default.nix:
```diff
@@ -18,6 +18,11 @@ stdenv.mkDerivation rec {
   ];
 
+  buildPhase = ''
+    make compiler
+    make compiler
```
Why twice?

NieDzejkob replied:
It seemed to me that the commit message has made this clear. Should I copy that explanation into a comment?

NieDzejkob replied:
Actually, I was wrong here. The build scripts used by Isabelle use `make compiler` twice, so that's what I tried at first, but now that I've checked, a single `make compiler` resolves the issue too.

I briefly browsed through PolyML's Makefile and I don't think there's anything that might make the result of a double `make compiler` differ from a single run.

SuperSandro2000 replied:
Yes please. I don't usually check extended commit messages.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -28374,6 +28381,8 @@ in
 
   monosat = callPackage ../applications/science/logic/monosat {};
 
+  naproche = callPackage ../applications/science/logic/naproche {};
```
```suggestion
  naproche = callPackage ../applications/science/logic/naproche { };
```

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 115339` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
5 packages built:
</summary>
<ul>
<li>hol</li>
<li>isabelle</li>
<li>naproche</li>
<li>polyml</li>
<li>tlaps</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
hol:
<details>


warning: stale-substitute
Stale substituteInPlace detected.
substituteStream(): WARNING: pattern '"/usr/bin/dot"' doesn't match anything in file 'tools/buildutils.sml'
See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/stale-substitute.md
warning: stale-substitute
Stale substituteInPlace detected.
substituteStream(): WARNING: pattern '"/usr/bin/dot"' doesn't match anything in file 'help/src-sml/DOT'
See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/stale-substitute.md
</details>
</details>

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 115339` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
5 packages built:
</summary>
<ul>
<li>hol</li>
<li>isabelle</li>
<li>naproche</li>
<li>polyml</li>
<li>tlaps</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
hol:
<details>


warning: stale-substitute
Stale substituteInPlace detected.
substituteStream(): WARNING: pattern '"/usr/bin/dot"' doesn't match anything in file 'tools/buildutils.sml'
See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/stale-substitute.md
warning: stale-substitute
Stale substituteInPlace detected.
substituteStream(): WARNING: pattern '"/usr/bin/dot"' doesn't match anything in file 'help/src-sml/DOT'
See: https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/stale-substitute.md
</details>
</details>

NieDzejkob commented:
I looked into the unnecessary substitution in the HOL package, but I don't understand how to trigger HOL's usage of dot, so I can't test any fixes.

SuperSandro2000 commented on pkgs/development/compilers/polyml/default.nix:
```diff
@@ -18,6 +18,10 @@ stdenv.mkDerivation rec {
   ];
 
+  buildPhase = ''
+    make compiler
```
Please use makeFlags.

NieDzejkob replied:
I don't think that's right, that would mean that the `compiler` target gets invoked during the install phase too.

SuperSandro2000 replied:
Why? Using ``makeFlags = [ "compiler" ];`` should do the same.

NieDzejkob replied:
`makeFlags = [ "compiler" ];` would do `make compiler` during `buildPhase` and `make compiler install` in the `installPhase`, which is wrong. However, I have found the `buildFlags` option, I will use it instead.

wizeman commented:
I've backported this to my local copy of the release-20.09 branch and it seems to be working for me, but I've only done some trivial testing.

I've ran into the following errors, though:

1. `/nix/store/xysfjmprg2w4yln2jni2qs81f6gmrk0k-isabelle-2021/Isabelle2021/lib/Tools/getenv: line 81: exec: perl: not found`

I've fixed this error with the following patch:
```diff
diff --git a/pkgs/applications/science/logic/isabelle/default.nix b/pkgs/applications/science/logic/isabelle/default.nix
index 397af8f32eb..258bcb2ccf9 100644
--- a/pkgs/applications/science/logic/isabelle/default.nix
+++ b/pkgs/applications/science/logic/isabelle/default.nix
@@ -92,6 +92,7 @@ stdenv.mkDerivation rec {
     mv $TMP/$dirname $out
     cd $out/$dirname
     bin/isabelle install $out/bin
+    substituteInPlace lib/Tools/getenv --replace perl ${perl}/bin/perl
   '';
 
   meta = with stdenv.lib; {
```

2. Errors creating directories when first starting `isabelle jedit`:
```
Failed to create directory: "/home/username/.isabelle/Isabelle2021/jedit/DockableWindowManager"
Failed to create directory: "/home/username/.isabelle/Isabelle2021/heaps/polyml-for-isabelle_x86_64-linux/log"
```

I've worked around this by creating the directories manually with `mkdir -p`.

All of these errors were also present in Isabelle 2020, though.

wizeman commented on pkgs/applications/science/logic/isabelle/default.nix:
```diff
@@ -1,24 +1,35 @@
-{ lib, stdenv, fetchurl, perl, perlPackages, makeWrapper, nettools, java, polyml, z3, rlwrap }:
+{ lib, stdenv, fetchurl, callPackage, runCommand, coreutils, perl, perlPackages
+, makeWrapper, nettools, java, polyml, z3, rlwrap, naproche }:
```
Are `perlPackages` and `makeWrapper` being used?

It doesn't seem that they are needed... `makeWrapper` is added as a dependency but then `makeWrapper`/`wrapProgram` is not called anywhere?

NieDzejkob replied:
Both `perlPackages` and `makeWrapper` are used here:

https://github.com/NixOS/nixpkgs/blob/c50f25092c20b7f8ae6bebd21d931963b5fc5b1f/pkgs/applications/science/logic/isabelle/default.nix#L96

NieDzejkob commented:
> `/nix/store/xysfjmprg2w4yln2jni2qs81f6gmrk0k-isabelle-2021/Isabelle2021/lib/Tools/getenv: line 81: exec: perl: not found`

Is that during `nix-build`, or...? I've got `perl` in `/run/current-system/sw/bin` (and in my tests it gets included even for a very barebones config), so I don't understand how it could trigger in your setup. (I do agree that we shouldn't rely on perl being in user's $PATH, just curious how you're testing to make this happen.)

As for the directory creation problem, it isn't something I've hit, either. I'd like to fix it though, is there anything unusual about your setup that could cause it? Are you using `home-manager` or something?

wizeman commented:
Thanks for pointing me to the location of `wrapProgram` and `perlPackages`, I had missed it because I had forgotten to backport the patch that added that line.

> Is that during `nix-build`, or...? I've got `perl` in `/run/current-system/sw/bin` (and in my tests it gets included even for a very barebones config), so I don't understand how it could trigger in your setup. (I do agree that we shouldn't rely on perl being in user's $PATH, just curious how you're testing to make this happen.)

The `perl` error happens when I first run `isabelle jedit`. I know this happens when I have a `$HOME/.isabelle` but not `$HOME/.isabelle/Isabelle2021`, but I don't know if that is needed.

I guess the main difference between our systems is that I *don't* have `perl` in `/run/current-system/sw/bin`.

I'm running NixOS 20.09, so I'm guessing that either your test environment somehow adds `perl` to `environment.systemPackages` or that the `master` branch added back `perl` to `environment.systemPackages` in some basic module.

> As for the directory creation problem, it isn't something I've hit, either. I'd like to fix it though, is there anything unusual about your setup that could cause it? Are you using `home-manager` or something?

Not that I can think of. Note that I already had `$HOME/.isabelle` and `Isabelle2016-1` and `Isabelle2017` subdirectories, but I didn't have a `Isabelle2021` subdirectory, but I don't know if that is related...

To be honest it seems like a weird error, why would creating a directory fail? Permissions and ownership seem perfectly fine.
The only thing I've noticed is that when I ran into those failures, Isabelle was trying to create directories without having created the parent directories first. But I still think it's weird that nobody else noticed this issue before...

NieDzejkob commented:
> Thanks for pointing me to the location of `wrapProgram` and `perlPackages`, I had missed it because I had forgotten to backport the patch that added that line.
> 
> > Is that during `nix-build`, or...? I've got `perl` in `/run/current-system/sw/bin` (and in my tests it gets included even for a very barebones config), so I don't understand how it could trigger in your setup. (I do agree that we shouldn't rely on perl being in user's $PATH, just curious how you're testing to make this happen.)
> 
> The `perl` error happens when I first run `isabelle jedit`. I know this happens when I have a `$HOME/.isabelle` but not `$HOME/.isabelle/Isabelle2021`, but I don't know if that is needed.
> 
> I guess the main difference between our systems is that I _don't_ have `perl` in `/run/current-system/sw/bin`.
> 
> I'm running NixOS 20.09, so I'm guessing that either your test environment somehow adds `perl` to `environment.systemPackages` or that the `master` branch added back `perl` to `environment.systemPackages` in some basic module.
> 
> > As for the directory creation problem, it isn't something I've hit, either. I'd like to fix it though, is there anything unusual about your setup that could cause it? Are you using `home-manager` or something?
> 
> Not that I can think of. Note that I already had `$HOME/.isabelle` and `Isabelle2016-1` and `Isabelle2017` subdirectories, but I didn't have a `Isabelle2021` subdirectory, but I don't know if that is related...
> 
> To be honest it seems like a weird error, why would creating a directory fail? Permissions and ownership seem perfectly fine.
> The only thing I've noticed is that when I ran into those failures, Isabelle was trying to create directories without having created the parent directories first. But I still think it's weird that nobody else noticed this issue before...

I just tried to reproduce this situation on my side and failed. If it's consistent, maybe you could try to capture an strace?

wizeman commented:
There's another difference between your PR and my test environment: I'm using Isabelle with openjdk11 instead of openjdk15, not sure if it's relevant.
I will try to capture an strace, thanks!

meithecatte committed:
isabelle: add perl to PATH

This makes 'isabelle getenv' work, as well as fixes an issue with
directory creation on first run of Isabelle.

wizeman commented:
Ok, so the strace didn't reveal much, it was `stat()`ing the `$HOME/.isabelle/Isabelle2021/jedit/DockableWindowManager` directory but it wasn't even trying to create it.

I searched for the place in the source code where the `DockableWindowManager` was being created and noticed it was calling a `make_directory()` function.

There seemed to be 2 `make_directory()` implementations, one in `src/Pure/System/isabelle_system.scala`, which looked fine, and another in `src/Pure/General/ssh.scala`, which was creating the directory via `perl`!

I don't know why on my system, it would call the SSH implementation, but sure enough, the strace revealed `perl` was being searched for in `$PATH` but wasn't being found. I don't know exactly why it was being searched for in `$PATH`, as no `exec*()` call was being executed, AFAICT.

Anyway, on a hunch that the problem was the missing `perl`, I've applied the following patch and it fixed the directory creating errors for me:
```diff
diff --git a/pkgs/applications/science/logic/isabelle/default.nix b/pkgs/applications/science/logic/isabelle/default.nix
index af20cb4b5b2..40d86ee6110 100644
--- a/pkgs/applications/science/logic/isabelle/default.nix
+++ b/pkgs/applications/science/logic/isabelle/default.nix
@@ -93,8 +93,7 @@ stdenv.mkDerivation rec {
     cd $out/$dirname
     bin/isabelle install $out/bin
 
-    substituteInPlace lib/Tools/getenv --replace perl ${perl}/bin/perl
-
+    wrapProgram $out/bin/isabelle --prefix PATH : "${lib.getBin perl}/bin"
     wrapProgram $out/$dirname/src/HOL/Tools/ATP/scripts/remote_atp --set PERL5LIB ${perlPackages.makeFullPerlPath [ perlPackages.LWP ]}
   '';
```

There was still a minor issue that the first time I ran `isabelle jedit` and then exited via the menu, the isabelle process seemed to hang after closing the window, but that only happened the first time -- it seems to be working fine now.

NieDzejkob commented:
What about other scripts in bin, such as `isabelle_java` or
`isabelle_scala_script`? AFAIK, the correct way to set up environment
variables for Isabelle is to edit `Isabelle2021/etc/settings`.

wizeman commented:
@NieDzejkob You're probably right about that, I don't know enough about Isabelle to tell either way...

SuperSandro2000 commented on pkgs/applications/science/logic/naproche/default.nix:
```diff
@@ -0,0 +1,31 @@
+  homepage = "https://github.com/naproche-community/naproche#readme";
+  description = "Write formal proofs in natural language and LaTeX";
+  maintainers = with lib.maintainers; [ NieDzejkob ];
+  license = lib.licenses.gpl3;
```
gpl3Only or gpl3Plus?

wizeman replied:
The source files don't seem to have a GPL header, so it's not clear. However, the `package.yaml` file says the license is `GPL-3`, so I would just go with `gpl3Only`.

wizeman commented:
Is there any chance we could get this merged anytime soon?

I just finished integrating Isabelle with Why3 and it took a lot of work... I was hoping to also submit my PR soon, but it depends on this PR being merged first.

Thanks!

wizeman commented:
BTW, I've been testing this PR continuously for more than a week and (after the perl dependency fix) it's been working great!
In particular, building the system heap during the Nix build is a great improvement from what we had before.

NieDzejkob commented:
All perl issues should now be fixed. Also corrected naproche's license to `gpl3Only`.

wizeman reviewed (approved):
Looks good to me, thanks!

stigtsp commented:
Hi there!

`perlPackages.PerlIOLayers: add perlPreHook for $LD override` is already in master as 496963f42819f4e20f5eea08da4c04b054aa4ce0 maybe a rebase is needed?

NieDzejkob commented:
> Hi there!
> 
> `perlPackages.PerlIOLayers: add perlPreHook for $LD override` is already in master as [496963f](https://github.com/NixOS/nixpkgs/commit/496963f42819f4e20f5eea08da4c04b054aa4ce0) maybe a rebase is needed?

Uh, sorry about that! Looks like a rebase has gone wrong in the first place to introduce this commit. Fixed now.

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -29013,11 +29013,18 @@ in
     });
 
-    java = openjdk11;
+    java = openjdk15;
```
```suggestion
    java = openjdk16;
```

I think we no longer have openjdk15 available.

NieDzejkob replied:
Okay, this is now fixed.

NieDzejkob commented:
Hmm, isabelle can be quite picky with the versions of its dependencies,
so I'd rather play it safe and use openjdk15. I'll take a look at
bringing it back from the depths of git history.

wizeman commented:
For what it's worth, I've been using Isabelle 2021 with openjdk11 and I haven't had any problems so far.

I'd imagine that if Isabelle lists openjdk15 as a dependency, then openjdk16 would probably also work fine since it's supposed to be backwards compatible?

NieDzejkob commented:
Sorry, it's been a while, but I've been busy. I've rebased this branch and modified it to use `openjdk16`. Everything seems to work fine.

SuperSandro2000 committed:
Update pkgs/applications/science/logic/isabelle/default.nix

SuperSandro2000 commented:
@NieDzejkob please resolve the merge conflict.

NieDzejkob commented:
#130583 is only a subset of the changes done here. For example, it disables the new Naproche component instead of packaging it as done here. I am currently not interested in resolving the conflicts.

jvanbruegge commented:
this can be closed now, all the things here are now in master
